staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -250,31 +314,49 @@ func (a *Authenticator) getToken(r *clientauthentication.Response) (string, erro
 	}
 
 	if err := cmd.Run(); err != nil {
-		return "", fmt.Errorf("exec: %v", err)
+		return fmt.Errorf("exec: %v", err)
 	}
 
 	_, gvk, err := codecs.UniversalDecoder(a.group).Decode(stdout.Bytes(), nil, cred)
 	if err != nil {
-		return "", fmt.Errorf("decode stdout: %v", err)
+		return fmt.Errorf("decode stdout (%q): %v", stdout.String(), err)
 	}
 	if gvk.Group != a.group.Group || gvk.Version != a.group.Version {
-		return "", fmt.Errorf("exec plugin is configured to use API version %s, plugin returned version %s",
+		return fmt.Errorf("exec plugin is configured to use API version %s, plugin returned version %s",
 			a.group, schema.GroupVersion{Group: gvk.Group, Version: gvk.Version})
 	}
 
 	if cred.Status == nil {
-		return "", fmt.Errorf("exec plugin didn't return a status field")
+		return fmt.Errorf("exec plugin didn't return a status field")
+	}
+	if cred.Status.Token == "" && cred.Status.ClientCertificateData == "" && cred.Status.ClientKeyData == "" {
+		return fmt.Errorf("exec plugin didn't return a token or cert/key pair")
 	}
-	if cred.Status.Token == "" {
-		return "", fmt.Errorf("exec plugin didn't return a token")
+	if (cred.Status.ClientCertificateData == "") != (cred.Status.ClientKeyData == "") {
+		return fmt.Errorf("exec plugin returned only certificate or key, not both")
 	}
 
 	if cred.Status.ExpirationTimestamp != nil {
 		a.exp = cred.Status.ExpirationTimestamp.Time
 	} else {
 		a.exp = time.Time{}
 	}
-	a.cachedToken = cred.Status.Token
 
-	return a.cachedToken, nil
+	oldCreds := a.cachedCreds
+	a.cachedCreds = &credentials{
+		token: cred.Status.Token,
+	}
+
+	// Avoid setting non-nil empty key/cert ([]byte{}) when plugin didn't return them.
+	if cred.Status.ClientKeyData != "" {
+		a.cachedCreds.key = []byte(cred.Status.ClientKeyData)
+	}
+	if cred.Status.ClientCertificateData != "" {
+		a.cachedCreds.cert = []byte(cred.Status.ClientCertificateData)
+	}
+
+	if a.onRotate != nil && !reflect.DeepEqual(oldCreds, a.cachedCreds) {
+		a.onRotate()
```

Comment:
We don't need to rotate if the Token changes but the cert doesn't. Maybe a TODO if it's hard to change right now.

Reply:
Done